Psychtoolbox/PsychBasic/DrawFormattedText2.m

```diff
@@ -0,0 +1,1329 @@
+% these strings with yPositionIsBaseline==true. Another difference is that
+% the returned textbounds bounding box includes the height of an empty line
+% at the end if the input string ended with a carriage return.
+% DrawFormattedText only moved (nx,ny) but did not include the empty line
+% in the bounding box. The empty line is also taken into account when
+% centering text
+%
+%
+% Further Notes:
+% 
+% Please note that while positioning and bounding boxes are pixel accurate
+% with the fonts tested during development, i cannot guarantee this is the
+% case with all fonts you throw at it. Also note that this function is not
+% made to be fast. It has a make draw from cache mode so that all
+% preprocessing is done only once and stored in a cache from which the text
+% can be drawn directly. Nonetheless, especially if you throw a lot of
+% formatting at it, this function may still not be fast in this mode. Use
+% with caution in timing critical paths (test and measure to know if its
+% fast enough). That said, with reasonable inputs it should manage to draw
+% text to screen well within the inter frame interval of most screens.
+%
+% The function employs clipping by default. Text lines that are detected as
+% laying completely outside the 'win'dow or optional 'winRect' will not be
+% drawn, but clipped away. This allows to draw multi-page text (multiple
+% screen heights) without too much loss of drawing speed. If you find the
+% clipping to interfere with text layout of exotic texts/fonts at exotic
+% sizes and formatting, you can define the global variable...
+% global ptb_drawformattedtext2_disableClipping;
+% ... and set it like this ...
+% ptb_drawformattedtext2_disableClipping = 1;
+% ... to disable the clipping.
+%
+% Regardless of the clipping setting, the optional 3rd return parameter
+% 'textbounds' always covers the complete text.
+%
+% See DrawFormattedText2Demo for a usage example.
+
+% TODO:
+% - justification. I have included an interface, but not implemented (or
+%   documented it in the help above) yet.
+% - Per word bounding boxes (optionally grow them to be proper AOIs for
+%   eye-tracking)
+
+% History:
+% 2015--2017    Written (DCN).
+
+
+global ptb_drawformattedtext2_disableClipping;
+global ptb_drawformattedtext2_padthresh;
+
+if isempty(ptb_drawformattedtext2_disableClipping)
+    % Text clipping on by default:
+    ptb_drawformattedtext2_disableClipping = 0;
+end
+% Boundinx boxes are always for the whole text, but text out of the screen
+% is culling/clipping. User can forcefully disable or enable this clipping.
+disableClip = (ptb_drawformattedtext2_disableClipping ~= -1) && ...
+              ((ptb_drawformattedtext2_disableClipping > 0));
+
+if isempty(ptb_drawformattedtext2_padthresh)
+    % Threshold for skipping of text justification is 33% by default:
+    ptb_drawformattedtext2_padthresh = 0.333;
+end
+padthresh = ptb_drawformattedtext2_padthresh;
+
+
+%% process key-value input

```

Comment:
Should get rid of those ^M s in some places, where line-feeds are somehow messed up.